src/box/lua/config/applier/roles.lua

```diff
@@ -1,16 +1,78 @@
 local log = require('internal.config.utils.log')
+local fiber = require('fiber')
 
-local last_loaded = {}
-local last_roles_ordered = {}
+local roles_state = {
+    -- Loaded roles, where the key is the role name and the value is the
+    -- role object.
+    last_loaded = {},
+
+    -- Ordered list of roles, where the order is determined by the
+    -- dependencies between roles.
+    last_roles_ordered = {},
+
+    -- The box.status watcher is used to call on_box_status callbacks.
+    box_status_watcher = nil,
+
+    -- Last values received from box.status watcher.
+    last_box_status_value = nil,
+
+    -- Last applied config.
+    last_config = nil,
+}
+
+-- Call all the 'on_box_status' callbacks from roles in order.
+local function call_on_box_status_callbacks(config, value)
+    for _, role_name in ipairs(roles_state.last_roles_ordered) do
+        local role = roles_state.last_loaded[role_name]
```

Comment:
I guess a reconfiguration may occur during the iteration, changing these values.

I would copy these two tables before the cycle.

Reply:
Thanks for noticing! Fixed.

Reply:
I see now: no need to copy, these tables are reassigned to the new ones each time and not modified. However, `last_loaded` could be changed, so my comment was partially relevant.

So, storing the references before the cycle is actually enough. Thanks!